kube: report errors back to user on every failure

tigrato opened the pull request:
This PR fixes an issue where the client could miss the exit error because the session was abruptly closed.

This PR adds coverage to ensure recorder failures force the session to be closed.

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
+
+		onErr(returnErr)
 		// call onFinished to emit the session.end and exec events.
+		// onFinish is never nil
```
nit:

```suggestion
		// onFinish is never nil.
```

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
+	onFinished, err := s.lockedSetupLaunch(request, eventPodMeta)
 	defer func() {
+		onErr := func(err error) {
+			s.setTerminationErr(err)
```
Should we check `err != nil` before running the code below, particularly the log statement?

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -547,14 +552,22 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
+			s.log.WithError(err).Warning("Executor failed while streaming.")
+		}
+
+		onErr(returnErr)
```
Suggestion: ditch the additional func and move the code here?

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
+		Component:    teleport.Component(teleport.ComponentSession, teleport.ComponentProxyKube),
+		// Session stream is using server context, not session context,
+		// to make sure that session is uploaded even after it is closed
+		Context: s.forwarder.ctx,
+	})
+	if err != nil {
+		return onFinish, trace.Wrap(err)
+	}
+
+	s.recorder = recorder
+	s.emitter = s.forwarder.cfg.Emitter
+
+	s.io.AddWriter(sessionRecorderID, recorder)
+
+	// If the identity is verified with an MFA device, we enabled MFA-based presence for the session.
+	if s.PresenceEnabled {
+		s.eventsWaiter.Add(1)
+		go func() {
+			defer s.eventsWaiter.Done()
+			ticker := time.NewTicker(PresenceVerifyInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					err := s.checkPresence()
+					if err != nil {
+						s.log.WithError(err).Error("Failed to check presence, closing session as a security measure")
+						err := s.Close()
+						if err != nil {
+							s.log.WithError(err).Error("Failed to close session")
+						}
+					}
+				case <-s.closeC:
+					return
+				}
+			}
+		}()
+	}
+	return onFinish, nil
```
Is this block a move or are there changes here?

tigrato replied:
Just a simple move.

codingllama replied:
I thought so as I was eyeballing it, but thanks for confirming.

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
+					err := s.checkPresence()
+					if err != nil {
+						s.log.WithError(err).Error("Failed to check presence, closing session as a security measure")
+						err := s.Close()
```
Should we exit the loop after the Close, or does this keep working somehow?

If it does keep working, please add a brief comment explaining it.

tigrato replied:
Nice catch. I just moved this code and didn't saw it

codingllama commented on lib/kube/proxy/sess.go:
```diff
@@ -921,7 +863,57 @@ func (s *session) lockedSetupLaunch(request *remoteCommandRequest, q url.Values,
+					err := s.checkPresence()
+					if err != nil {
+						s.log.WithError(err).Error("Failed to check presence, closing session as a security measure")
+						err := s.Close()
+						if err != nil {
```
nit: Use the smaller scope, give it a space.

```suggestion

						if err := s.Close(); err != nil {
```

codingllama commented on lib/kube/proxy/sess_test.go:
```diff
@@ -73,99 +72,140 @@ func TestSessionEndError(t *testing.T) {
+			name:         "interactive with recording error",
+			interactive:  true,
+			recordingErr: recordingErr,
+		},
```
Should we test non-interactive with recording error?

tigrato replied:
non interactive sessions aren't recorded

codingllama commented on lib/kube/proxy/sess_test.go:
```diff
@@ -73,99 +72,140 @@ func TestSessionEndError(t *testing.T) {
+					assert.Truef(t, hasSessionExecEvent, "session exec event not found in audit log")
+				}, 10*time.Second, 1*time.Second)
+			} else {
+				require.Never(t, func() bool {
```
Do we need the periodic checking or is this equivalent to sleep(1s) -> check?

Could we do this in a way that doesn't take 1s?

tigrato replied:
unfortunately no. They are async and we need to wait for collection. The test takes 5s without this

codingllama commented on lib/kube/proxy/sess.go:
What is the change here that fixes the PR title? I'm not super familiar with lib/kube, so it's not super clear to me.

tigrato replied:
This section https://github.com/gravitational/teleport/pull/42154/files#diff-a98dd59a36206bb300939248013dd2e1b713077750f5d7b7c0c7e8b2d7fa443bR555-R564

tigrato replied:
If you want, I can split everything but takes a bit of time

codingllama replied:
No need to split anything, I just wanted to clarify.

The lockedSetupLaunch() error wasn't handled by an `onErr`, right?

tigrato committed:
kube: report errors back to user on every failure

This PR fixes an issue where the client could miss the exit error
because the session was abruptly closed.

This PR adds coverage to ensure recorder failures force the session to
be closed.

Signed-off-by: Tiago Silva <tiago.silva@goteleport.com>

AntonAM commented on lib/kube/proxy/sess.go:
```diff
@@ -547,14 +551,20 @@ func (s *session) launch(isEphemeralCont bool) (returnErr error) {
+			s.log.WithError(returnErr).Warning("Executor failed while streaming.")
+		}
 		// call onFinished to emit the session.end and exec events.
+		// onFinish is never nil.
```
```suggestion
		// onFinished is never nil.
```

tigrato committed:
Update sess.go

Co-authored-by: Anton Miniailo <anton@goteleport.com>
